miden-crypto/src/merkle/mmr/forest.rs

```diff
@@ -0,0 +1,458 @@
+
+        let idx = nodes + open_trees - right_subtree_count;
+
+        InOrderIndex::new(idx.try_into().unwrap())
+    }
+
+    /// Returns the in-order index of the rightmost element (the smallest tree).
+    pub fn rightmost_in_order_index(&self) -> InOrderIndex {
+        // Count total size of all trees in the forest.
+        let nodes = self.num_nodes();
+
+        // Add the count for the parent nodes that separate each tree. These are allocated but
+        // currently empty, and correspond to the nodes that will be used once the trees are merged.
+        let open_trees = self.num_trees() - 1;
+
+        let idx = nodes + open_trees;
+
+        InOrderIndex::new(idx.try_into().unwrap())
+    }
+
+    /// Given a leaf index in the current forest, return the tree number responsible for the
+    /// leaf.
+    ///
+    /// Note:
+    /// The result is a tree position `p`, it has the following interpretations:
+    /// - `p+1` is the depth of the tree.
+    /// - Because the root element is not part of the proof, `p` is the length of the authentication
+    ///   path.
+    /// - `2^p` is equal to the number of leaves in this particular tree.
+    /// - And `2^(p+1)-1` corresponds to the size of the tree.
+    ///
+    /// For example, given a forest with 6 leaves whose forest is `0b110`:
+    /// ```text
+    ///       __ tree 2 __
+    ///      /            \
+    ///    ____          ____         _ tree 1 _
+    ///   /    \        /    \       /          \
+    ///  0      1      2      3     4            5
+    /// ```
+    ///
+    /// Leaf indices `0..=3` are in the tree at index 2 and leaf indices `4..=5` are in the tree at
+    /// index 1.
+    pub fn leaf_to_corresponding_tree(self, leaf_idx: usize) -> Option<u32> {
+        let forest = self.0;
+
+        if leaf_idx >= forest {
+            None
+        } else {
+            // - each bit in the forest is a unique tree and the bit position is its power-of-two
+            //   size
+            // - each tree is associated to a consecutive range of positions equal to its size from
+            //   left-to-right
+            // - this means the first tree owns from `0` up to the `2^k_0` first positions, where
+            //   `k_0` is the highest set bit position, the second tree from `2^k_0 + 1` up to
+            //   `2^k_1` where `k_1` is the second highest bit, so on.
+            // - this means the highest bits work as a category marker, and the position is owned by
+            //   the first tree which doesn't share a high bit with the position
+            let before = forest & leaf_idx;
+            let after = forest ^ before;
+            let tree_idx = after.ilog2();
+
+            Some(tree_idx)
+        }
+    }
+
+    /// Given a leaf index in the current forest, return the lead index in the tree to which
+    /// the leaf belongs..
+    pub(super) fn leaf_relative_position(self, leaf_idx: usize) -> Option<usize> {
```

Comment:
What does "lead index" mean here? Also, I think there is an extra period at the ned of the sentence.